### ISSUE TITLE
[BUG] Incorrect modification date in savegame internal files

### ISSUE THREAD

**Issue by itaranto** [ZDoom/gzdoom#2306]
### GZDoom version

GZDoom version g4.11.3-m

### Which game are you running with GZDoom?

Doom 2

### What Operating System are you using?

Linux x86_64

### Please describe your specific OS version

OS: Arch Linux, Linux

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Files inside savegames (.zds files) have incorrect modification dates.

For example:

```
$ unzip -v auto00.zds
Archive:  auto00.zds
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
   31779  Stored    31779   0% 1980-09-13 10:60 4daafd24  savepic.png
     316  Defl:X      189  40% 1980-09-13 10:60 a77286f8  info.json
   79736  Defl:X    37383  53% 1980-09-13 10:60 1ad9f249  globals.json
  201511  Defl:X    16050  92% 1980-09-13 10:60 10d03dc2  map06.map.json
--------          -------  ---                            -------
  313342            85401  73%                            4 files
 ```
 
 And another:
 
 ```
$ unzip -v save00.zds
Archive:  save00.zds.bak
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
   42468  Stored    42468   0% 2073-03-19 10:60 c0c5abfb  savepic.png
     289  Defl:X      183  37% 2073-03-19 10:60 42d5b750  info.json
   79015  Defl:X    37133  53% 2073-03-19 10:60 d69a6312  globals.json
  102892  Defl:X    12575  88% 2073-03-19 10:60 412aadef  map01.map.json
--------          -------  ---                            -------
  224664            92359  59%                            4 files
```

The second example shows a future date (2073-...) , this is what got me to find the bug in the first place. I was trying to update the `info.json` file inside the Zip file like this:

 ```
$ zip -v -f save00.zds info.json
zip diagnostic: info.json up to date
 ```
 
The `zip` command output is correct, it doesn't actually update the file since the existing modification date is higher than the modified `info.json` file.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. `cd ~/.config/gzdoom/savegames/doom.id.doom2.commercial`
2. Print a savegame's information: `unzip -v save00.zds`
3. Observe the files' modification dates are really old (like 1981 or something) or in the future (like 2073).

**Comment by coelckers** [comment]
This one's ancient, unfortunately I have never found out what is wrong about the code that generates the dates. :(

**Comment by itaranto** [comment]
> This one's ancient, unfortunately I have never found out what is wrong about the code that generates the dates. :(

Well I took a look at the source code and I think I found the issue :)

TLDR, the code seems to be reading the mod date as time and the mod time as date, see below for the explanation.

In the function `WriteZip` in file `file_zip.cpp` there's the code that writes all the file entries with their headers.

I saw you obtain the time using some standard C functions and then you convert it two "DOS time" using the `time_to_dos` function. This is because the Zip format uses MS-DOS time and date formats.

So I went ahead and printed both the "C time" and the "DOS time" to see if there's any difference, this is what I did:

```diff
diff --git a/src/common/filesystem/source/file_zip.cpp b/src/common/filesystem/source/file_zip.cpp
index 28c2dedec..3802fa5bf 100644
--- a/src/common/filesystem/source/file_zip.cpp
+++ b/src/common/filesystem/source/file_zip.cpp
@@ -40,6 +40,7 @@
 #include "ancientzip.h"
 #include "fs_findfile.h"
 #include "fs_swap.h"
+#include <stdio.h>
 
 namespace FileSys {
 	using namespace byteswap;
@@ -664,6 +665,20 @@ int AppendCentralDirectory(FileWriter *zip_file, const FCompressedBuffer &conten
 	return 0;
 }
 
+void print_dos_time(uint16_t value) {
+	auto hours = value >> 11;
+	auto minutes = (value >> 5) & 0x3F;
+	auto seconds = (value & 0x1F) * 2;
+	printf("DOS last mod file time: %d:%d:%d\n", hours, minutes, seconds);
+}
+
+void print_dos_date(uint16_t value) {
+	auto day = value & 0x1F;
+	auto month = (value >> 5) & 0x0F;
+	auto year = ((value >> 9) & 0x7F) + 1980;
+	printf("DOS last mod file date: %d:%d:%d\n", day, month, year);
+}
+
 bool WriteZip(const char* filename, const FCompressedBuffer* content, size_t contentcount)
 {
 	// try to determine local time
@@ -671,8 +686,14 @@ bool WriteZip(const char* filename, const FCompressedBuffer* content, size_t con
 	time_t ttime;
 	ttime = time(nullptr);
 	ltime = localtime(&ttime);
+
+	printf("The current time is %s(%jd seconds since the Epoch)\n", asctime(ltime), (intmax_t)ttime);
+
 	auto dostime = time_to_dos(ltime);
 
+	print_dos_date(dostime.first);
+	print_dos_time(dostime.second);
+
 	TArray<int> positions;
 
 	auto f = FileWriter::Open(filename);
```

(BTW I just searched in the internet how to print DOS dates and times because I didn't want to figure out the bitwise logic).

So the results were, surprisingly correct:

```
The current time is Sun Dec 17 17:51:59 2023======================================================================....]
(1702846319 seconds since the Epoch)
DOS last mod file date: 17:12:2023
DOS last mod file time: 17:51:58
Game saved.
```

Although the date and time were still wrong inside the Zip file.

Then I thought, what if swap `dostime.first` and `dostime.second` ?

 I mean:

```c
	print_dos_date(dostime.second);
	print_dos_time(dostime.first);
```

and this is what I got:

```
The current time is Sun Dec 17 17:55:38 2023======================================================================....]
(1702846538 seconds since the Epoch)
DOS last mod file date: 19:7:2051
DOS last mod file time: 10:60:34
Game saved.
```

Which is exactly what I see on the actual file.

So the code that's reading the Zip file, seems to be reading the fields in the wrong order. I believe that code is in the `CheckZip` function.

**Comment by itaranto** [comment]
Well, it seems the fix is simple:

By swapping `first` with `second` in the `time_to_dos` function, the issue seems to be fixed.

**Comment by coelckers** [comment]
Would you mind making a PR for this so the commit history gives you proper credit? Otherwise I'll apply the fix later today.

**Comment by itaranto** [comment]
> Would you mind making a PR for this so the commit history gives you proper credit? Otherwise I'll apply the fix later today.

Sure, I'll open a PR later today.

**Comment by itaranto** [comment]
Here's the PR: https://github.com/ZDoom/gzdoom/pull/2311
